contrib/mesos/pkg/executor/executor.go

```diff
@@ -1044,3 +1002,220 @@ func executorInfoToAnnotations(ei *mesos.ExecutorInfo) (annotations map[string]s
 
 	return
 }
+
+// implements txTaskOperator, uses the executor state lock to handle concurrent invocation
+func (k *Executor) doForTask(taskId string, f func(*kuberTask, *api.Pod) bool) (changed bool, err error) {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	// exclude tasks which are already deleted from our task registry
+	task := k.tasks[taskId]
+	if task == nil {
+		// the pod has completed the launch-task-binding phase because it's been annotated with
+		// the task-id, but we don't have a record of it; it's best to let the scheduler reconcile.
+		// it's also possible that our update queue is backed up and hasn't caught up with the state
+		// of the world yet.
+
+		// TODO(jdef) should we hint to the scheduler (via TASK_FAILED, reason=PodRefersToUnknownTask)?
+
+		err = fmt.Errorf("task %s not found", taskId)
+		return
+	}
+
+	oldPod := k.pods[task.podName]
+	if oldPod == nil {
+		// should never happen because:
+		// (a) the update pod record has already passed through the binding phase in launchTasks()
+		// (b) all remaining updates to executor.{pods,tasks} are sync'd in unison
+		err = fmt.Errorf("internal state error: pod not found for task %s", taskId)
+		return
+	}
+
+	changed = f(task, oldPod)
+	if changed {
+		k.sendPodsSnapshot()
+	}
+	return
+}
+
+type podUpdateProcessor struct {
```

Comment:
will do

On Wed, Dec 9, 2015 at 11:43 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/executor.go
> <https://github.com/kubernetes/kubernetes/pull/17917#discussion_r47116184>
> :
>
> > +	if oldPod == nil {
> > +		// should never happen because:
> > +		// (a) the update pod record has already passed through the binding phase in launchTasks()
> > +		// (b) all remaining updates to executor.{pods,tasks} are sync'd in unison
> > +		err = fmt.Errorf("internal state error: pod not found for task %s", taskId)
> > +		return
> > +	}
> > +
> > +	changed = f(task, oldPod)
> > +	if changed {
> > +		k.sendPodsSnapshot()
> > +	}
> > +	return
> > +}
> > +
> > +type podUpdateProcessor struct {
>
> Please move thin into its own module. It's far too complex for executor.go.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17917/files#r47116184>.
>